File: May-LeetCoding-Challenge/27-Possible-Bipartition/Possible-Bipartition.cpp
class Solution {
public:
    vector<int> color;
    vector<vector<int>> G;
    
    // Return true if the connected component of u is a bipartite graph
    bool bipartite(int u) {
        for (int i = 0; i < G[u].size(); i++) {
            int v = G[u][i];
            if (color[v] == color[u]) return false;
            if (!color[v]) {
                color[v] = 3 - color[u];
                if (!bipartite(v)) return false;
            }
        }
        return true;
    }
    
    bool possibleBipartition(int N, vector<vector<int>>& dislikes) {
        color.resize(N+1);
        G.resize(N+1);

        
        for (int i = 0; i < dislikes.size(); i++) {
            G[dislikes[i][0]].push_back(dislikes[i][1]);
            G[dislikes[i][1]].push_back(dislikes[i][0]);
        }
        
        for (int i = 1; i <= N; i++)
            if (!color[i]){
                color[i] = 1;
                if (!bipartite(i)) return false;
            }
        return true;
        
    }
};

File: May-LeetCoding-Challenge/27-Possible-Bipartition/Possible-Bipartition.java
import java.util.HashMap;

class Solution {
    public boolean possibleBipartition(int N, int[][] dislikes) {
        /**
         * Build a graph, try to color each node
         * 0, If the node hasn't been colored, => color to 1
         * 1, color its dislikes nodes to - color => color to -1
         * 2, conflct return false
         */
        
        Map<Integer, Set<Integer>> g = new HashMap<>();

        // build graph
        for (int[] d : dislikes) {
            int a = d[0];
            int b = d[1];
            g.putIfAbsent(a, new HashSet<Integer>());
            g.putIfAbsent(b, new HashSet<Integer>());
            g.get(a).add(b);
            g.get(b).add(a);
        }

        int[] colors = new int[N + 1];
        for (int i = 1; i <= N; i++) {
            // the node is not colored and there is a conflict
            if (colors[i] == 0 && !dfs(colors, 1, i, g)) return false;
        }

        return true;
    }

    private boolean dfs (int[] colors, int color, int node, Map<Integer, Set<Integer>> g) {
        // The node is already colored
        if (colors[node] != 0) {
            return colors[node] == color;
        }

        colors[node] = color;
        if (g.get(node) == null) return true;
        for (int next : g.get(node)) {
            if (!dfs(colors, -color, next, g)) return false;
        }

        return true;
    }
}